src/Analyzer/Passes/CrossToInnerJoinPass.cpp

```diff
@@ -0,0 +1,266 @@
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/ColumnNode.h>
+
+#include <Functions/FunctionFactory.h>
+#include <Functions/IFunction.h>
+
+#include <Common/logger_useful.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_QUERY;
+}
+
+namespace
+{
+
+void exctractJoinConditions(const QueryTreeNodePtr & node, QueryTreeNodes & equi_conditions, QueryTreeNodes & other)
+{
+    auto * func = node->as<FunctionNode>();
+    if (!func)
+    {
+        other.push_back(node);
+        return;
+    }
+
+    const auto & args = func->getArguments().getNodes();
+
+    if (args.size() == 2 && func->getFunctionName() == "equals")
+    {
+        equi_conditions.push_back(node);
+    }
+    else if (func->getFunctionName() == "and")
+    {
+        for (const auto & arg : args)
+            exctractJoinConditions(arg, equi_conditions, other);
+    }
+    else
+    {
+        other.push_back(node);
+    }
+}
+
+const QueryTreeNodePtr & getEquiArgument(const QueryTreeNodePtr & cond, size_t index)
+{
+    const auto * func = cond->as<FunctionNode>();
+    chassert(func && func->getFunctionName() == "equals" && func->getArguments().getNodes().size() == 2);
+    return func->getArguments().getNodes()[index];
+}
+
+
+/// Check that node has only one source and return it.
+/// {_, false} - multiple sources
+/// {nullptr, true} - no sources
+/// {source, true} - single source
+std::pair<const IQueryTreeNode *, bool> getExpressionSource(const QueryTreeNodePtr & node)
+{
+    if (const auto * column = node->as<ColumnNode>())
+    {
+        auto source = column->getColumnSourceOrNull();
+        if (!source)
+            return {nullptr, false};
+        return {source.get(), true};
+    }
+
+    if (const auto * func = node->as<FunctionNode>())
+    {
+        const IQueryTreeNode * source = nullptr;
+        const auto & args = func->getArguments().getNodes();
+        for (const auto & arg : args)
+        {
+            auto [arg_source, is_ok] = getExpressionSource(arg);
+            if (!is_ok)
+                return {nullptr, false};
+
+            if (!source)
+                source = arg_source;
+            else if (arg_source && !source->isEqual(*arg_source))
+                return {nullptr, false};
+        }
+        return {source, true};
+
+    }
+
+    if (node->as<ConstantNode>())
+        return {nullptr, true};
+
+    return {nullptr, false};
+}
+
+bool findInTableExpression(const IQueryTreeNode * source, const QueryTreeNodePtr & table_expression)
+{
+    if (!source)
+        return true;
+
+    if (source->isEqual(*table_expression))
+        return true;
+
+    if (const auto * join_node = table_expression->as<JoinNode>())
+    {
+        return findInTableExpression(source, join_node->getLeftTableExpression())
+            || findInTableExpression(source, join_node->getRightTableExpression());
+    }
+
+    if (const auto * query_node = table_expression->as<QueryNode>())
```

Comment:
I suspect that this need to be removed.
Because we call this function
```
auto left_src = getExpressionSource(getEquiArgument(cond, 0));
auto right_src = getExpressionSource(getEquiArgument(cond, 1));
if (left_src.second && right_src.second && left_src.first && right_src.first)
{
    bool can_join_on = (findInTableExpression(left_src.first, left_table) && findInTableExpression(right_src.first, right_table))
    || (findInTableExpression(left_src.first, right_table) && findInTableExpression(right_src.first, left_table));
```
And if there will be query like:
```
SELECT id FROM test_table, (SELECT 1 AS a) AS subquery WHERE condition.
```
We should not dive into `subquery` table expressions.